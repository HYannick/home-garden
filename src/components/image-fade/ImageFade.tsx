/** @jsx jsx */
import React, { useEffect, useState } from 'react';
import styled from '@emotion/styled';
import { jsx } from '@emotion/core';

interface ContainerProps {
  placeholder: string
}

const Container = styled('div')<ContainerProps>`
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  background-size: cover;
  background-color: ${({ placeholder }) => placeholder};
`;

interface MediaProps {
  fit?: boolean,
  imageLoaded: boolean,
  imageCached: boolean,
  transitionDuration?: number
}

export const Media = styled('img')<MediaProps>`
  width: ${({ fit }) => fit ? '100%' : 'initial'};
  height: 100%;
  object-fit:cover;
  opacity: ${({ imageLoaded }) => imageLoaded ? 1 : 0};
  transition: ${({ transitionDuration }) => `opacity ${transitionDuration}s`};
`;

export const ImageFade: React.FC<any> = ({ alt, src, placeholder = '#fff', transitionDuration = 0.3, fit = true }) => {
  const [imageLoaded, setImageLoaded] = useState(false);
  const [imageCached, setImageCached] = useState(false);

  useEffect(() => {
    const image = new Image();
    image.src = src;
    setImageCached(image.complete);
  }, [src]);

  const imageLoadHandler = () => setImageLoaded(true);


  return (
    <Container placeholder={placeholder}>
      <Media
        fit={fit}
        alt={alt}
        src={src}
        imageLoaded={imageLoaded}
        imageCached={imageCached}
        transitionDuration={transitionDuration}
        onLoad={imageLoadHandler}
      />
    </Container>
  );
};

export default ImageFade;
