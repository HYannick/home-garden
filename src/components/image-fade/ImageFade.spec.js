import React from 'react';
import { shallow } from 'enzyme/build';
import toJson from 'enzyme-to-json';
import ImageFade, { Media } from './ImageFade';

describe('Image Fade', () => {
  const IMAGE_PATH = 'https://images.unsplash.com/photo-1498955472675-532cdee9d6b4?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2151&q=80';
  const imgProps = {
    imageCached: true,
    imageLoaded: true,
    setImageCached: jest.fn(),
    setImageLoaded: jest.fn()
  };

  it('Should match snapshot when using Image Fade', () => {
    const wrapper = shallow(<ImageFade
      placeholder="linear-gradient(102deg, #8fc6dc, #8fd6dc)"
      src={IMAGE_PATH}
      {...imgProps}
      transitionDuration={0.7}
      alt="pic"
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('Should match snapshot when using Image Fade with FitMode', () => {
    const wrapper = shallow(<ImageFade
      placeholder="linear-gradient(102deg, #8fc6dc, #8fd6dc)"
      fit
      src={IMAGE_PATH}
      {...imgProps}
      transitionDuration={0.7}
      alt="pic"
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('Should have an image with the right src', () => {
    const wrapper = shallow(<ImageFade
      placeholder="linear-gradient(102deg, #8fc6dc, #8fd6dc)"
      src={IMAGE_PATH}
      {...imgProps}
      transitionDuration={0.7}
      alt="pic"
    />);

    expect(wrapper.find(Media).prop('src')).toEqual(IMAGE_PATH);
  });
});
