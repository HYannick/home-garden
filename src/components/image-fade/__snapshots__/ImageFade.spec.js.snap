// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Fade Should match snapshot when using Image Fade 1`] = `
<Styled(div)
  placeholder="linear-gradient(102deg, #8fc6dc, #8fd6dc)"
>
  <Styled(img)
    alt="pic"
    fit={true}
    imageCached={false}
    imageLoaded={false}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1498955472675-532cdee9d6b4?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2151&q=80"
    transitionDuration={0.7}
  />
</Styled(div)>
`;

exports[`Image Fade Should match snapshot when using Image Fade with FitMode 1`] = `
<Styled(div)
  placeholder="linear-gradient(102deg, #8fc6dc, #8fd6dc)"
>
  <Styled(img)
    alt="pic"
    fit={true}
    imageCached={false}
    imageLoaded={false}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1498955472675-532cdee9d6b4?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2151&q=80"
    transitionDuration={0.7}
  />
</Styled(div)>
`;
